src/utils/sources-tree/treeOrder.js

```diff
@@ -0,0 +1,131 @@
+// @flow
+
+import { parse } from "url";
+
+import { nodeHasChildren } from "./utils";
+
+import type { Node } from "./types";
+
+/*
+ * Gets domain from url (without www prefix)
+ */
+export function getDomain(url: string): ?string {
+  const { host } = parse(url);
+  if (!host) {
+    return null;
+  }
+  return host.startsWith("www.") ? host.substr("www.".length) : host;
+}
+
+/*
+ * Checks if node name matches debugger host/domain.
+ */
+function isExactDomainMatch(part: string, debuggeeHost: string): boolean {
+  return part.startsWith("www.")
+    ? part.substr("www.".length) === debuggeeHost
+    : part === debuggeeHost;
+}
+
+/*
+ * Function to assist with node search for a defined sorted order, see e.g.
+ * `createTreeNodeMatcher`. Returns negative number if the node
+ * stands earlier in sorting order, positive number if the node stands later
+ * in sorting order, or zero if the node is found.
+ */
+export type FindNodeInContentsMatcher = (node: Node) => number;
+
+/*
+ * Performs a binary search to insert a node into contents. Returns positive
+ * number, index of the found child, or negative number, which can be used
+ * to calculate a position where a new node can be inserted (`-index - 1`).
+ * The matcher is a function that returns result of comparision of a node with
+ * lookup value.
+ */
+export function findNodeInContents(
+  tree: Node,
+  matcher: FindNodeInContentsMatcher
+) {
+  const { contents } = tree;
+  if (contents.length === 0) {
+    return { found: false, index: 0 };
+  }
+  let left = 0;
+  let right = contents.length - 1;
+  while (left < right) {
+    const middle = Math.floor((left + right) / 2);
+    if (matcher(contents[middle]) < 0) {
+      left = middle + 1;
+    } else {
+      right = middle;
+    }
+  }
+  const result = matcher(contents[left]);
+  if (result === 0) {
+    return { found: true, index: left };
+  }
+  return { found: false, index: result > 0 ? left : left + 1 };
```

Comment:
thanks for clarifying why you were doing what you were doing before! The garbage collection might be an issue, true. I hadn't thought of that, and you are totally right

Perhaps a way to make the code readable and more memory efficient (trying to balance things :sweat_smile:), we could keep the code you had (with a comment about what it was doing, a more explicit comment)

then where the value is used, have a wrapper function that translates the value? 

ie)

```javascript
export function findNodeInContents(
  // ...
  // returns a negative value for a new node, offset by one
  // returns a positive value for an existing node
  return -(result > 0 ? left : left + 1 ) - 1;
```

with helpers
```javascript
function childFound(child) {
  return child > -1
} 

function determineIndex(child) {
  // return a valid index whether the child is new (an therefore a negative number off by one)
  // or if the child is an existing index
  return child > -1 ? child : -child - 1;
}
```


then where it is used:
```javascript
if (!childFound(childIndex)) {
  return createNodeInTree(part, path, subTree, determineIndex(childIndex));
}
```

Reply:
I would keep that as is for now for readability. We can re-open the issue later. It's hard to measure atm -- it's impossible to run it with 10,000 or 30,000 load.